### PR TITLE
[feat/aut169] Add getID, getName and getDescription to SequenceProcessor

### PR DESCRIPTION
**Pull request by ChristopherAbram**
✅ Add `IProcessorDescription` interface,
✅ Implement `IProcessorDescription` interface in `SequenceProcessor` classes,
✅ Add unit tests for relevant `SequenceProcessor`'s getters.

📌 Disclaimer:
- Assigned IDs are now set to `-1`, this probably needs an update,
- Descriptions might be not descriptive enough - I copied class docs to the string,
- Since I created `IProcessorDescription` interface, I can also update other Processor classes with this interface (out of the scope of this ticket)
- Maybe location of `IProcessorDescription` can be under new directory `interface` or `iface`:
image

### CODE REVIEW

**Comment by bertrandboudaud** [comment]
Yes, would be nice to use this interface for other processors. I'm not sure it's so out of scope of the ticket. since you introduce an interface, it may impact other class. is it lot of work? i feel if it's very easy you could go ahead. if not create a story once it is merged.

about the interface directory, yes iface is nice. but if we do so, all the interfaces should go there. would be better to put them all at the same time. perhaps it would be better to stay like you have done and create a story about that ("move all interfaces in dedicated directory").

File: src/smartpeak/include/SmartPeak/core/SequenceProcessor.h
Comment (bertrandboudaud) [inline src/smartpeak/include/SmartPeak/core/SequenceProcessor.h]: "override" would be welcome for all these getID() getName() and getDescription()

File: src/smartpeak/include/SmartPeak/core/IProcessorDescription.h
Comment (bertrandboudaud) [inline src/smartpeak/include/SmartPeak/core/IProcessorDescription.h]: This declaration can be usefull, but I would prefer a team agreement and to have a check on all code to use this pattern in a dedicated story to have consistent code. my concern is to have this pattern use only on some parts of code and not on others. to me explicit shared_ptr is nice because you can also have unique_ptr the declaration my hide this.

**Review by bertrandboudaud — CHANGES_REQUESTED** [review]

File: src/smartpeak/include/SmartPeak/core/SequenceProcessor.h
Comment (ChristopherAbram) [reply src/smartpeak/include/SmartPeak/core/SequenceProcessor.h]: Good point, `override` well describe developer's intention, however its optional

File: src/smartpeak/include/SmartPeak/core/IProcessorDescription.h
Comment (ChristopherAbram) [reply src/smartpeak/include/SmartPeak/core/IProcessorDescription.h]: Agree with your concerns, its a good practice to be consistent and keep same `coding conventions` everywhere in the project. We can have some discussion later during sprint review/retro.

**Comment by ChristopherAbram** [comment]
Should not be much work to add `IProcessorDescription` interface to another `Processor` classes. 
I am not sure however what you mean by moving all interface to one directory - at least if it comes to processor classes, I can see plenty of abstract classes, but not many classes that have only pure virtual functions. Moreover, these classes have very close `process()` function signatures, but still different. I think such work of defining interfaces for these classes would be, as you mentioned, another task, refactoring of code base

**Comment by bertrandboudaud** [comment]
I haven't check if there are other interfaces - sorry. If there is no other interfaces, you can go create the iface directory and put yours inside. I just would like (it's my point of view) not to have some interfaces in one directory, others (if they exist) in another.

**Review by bertrandboudaud — APPROVED** [review]

File: src/smartpeak/include/SmartPeak/iface/IProcessorDescription.h
Comment (dmccloskey) [inline src/smartpeak/include/SmartPeak/iface/IProcessorDescription.h]: Can you change to two spaces instead of tabs?

File: src/smartpeak/include/SmartPeak/iface/IProcessorDescription.h
Comment (dmccloskey) [inline src/smartpeak/include/SmartPeak/iface/IProcessorDescription.h]: Is this typdef used anywhere or can it be removed?

**Review by dmccloskey — CHANGES_REQUESTED** [review]
Interesting choice to implement an interface pattern for the processor classes.

If you don't mind taking care of my small comments I think we will be good to go once the CI is working.

**Review by dmccloskey — APPROVED** [review]

**Pull request merged**
